### PR TITLE
Fix bug in patching args in multiple runs

### PR DESCRIPTION
**Pull request by rbramand-xilinx**
#### Problem solved by the commit
When kernel is run multiple times and in some of the runs if arg is updated using set_arg the present code patches the arg address on top of existing already patched value this will result in test failures. This PR fixes this issue

#### How problem was solved, alternative solutions (if any) and why they were rejected
Storing BD data ptr entries in patch info and marking it dirty after first patching.
When second or subsequent run is made with set_arg the original BD data ptr values stored are restored and we patch on top of it which is proper.

#### Risks (if any) associated the changes in the commit
Moderate as this code changes xrt code critical path

#### What has been tested and how, request additional testing if necessary
Tested xrt elf flow test on aie2p by calling set_arg in multiple kernel runs and test passes
Also tested vadd application on aie2ps by modifying host application accordingly.
Thorough testing is needed to make sure this change doesn't break all the existing flows

#### Documentation impact (if any)
NA

### CODE REVIEW

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 1]
```diff
@@ -137,10 +137,13 @@ struct patcher
   buf_type m_buf_type = buf_type::ctrltext;
   symbol_type m_symbol_type = symbol_type::shim_dma_48;
 
+  static constexpr size_t num_bd_data_ptrs = 8; // max number of bd ptrs accessed while patching
   struct patch_info {
```
Comment (sonals) [inline src/runtime_src/core/common/api/xrt_module.cpp:141]: Can we add a comment on why **8** and make this a file constant?

Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/xrt_module.cpp]: Made changes accordingly Sonal, Thanks

**Review by chvamshi-xilinx — APPROVED** [review]
I know, this file is not following other XRT user space style of coding, Please update name of variable to lower case. "max_bd_words"

**Review by larry9523 — COMMENTED** [review]
This fix introduces a new dependency between xrt_core and AIE architecture. Ideally, all patching knowledges should come from ELF.

But given the plan that we will move all the patching service out of XRT to aiebu, I think this temporal solution is OK.

**Review by sonals — APPROVED** [review]

**Comment by larry9523** [comment]
@rbramand-xilinx  please provide a list of tests we have done before we are confident enough to merge this PR

**Comment by rbramand-xilinx** [comment]
> @rbramand-xilinx please provide a list of tests we have done before we are confident enough to merge this PR

Hi @larry9523 , I have tested xrt flow test on aie2p and vadd test on aie2ps by modifying host application and test passes on both of them.

**Comment by rbramand-xilinx** [comment]
> > @rbramand-xilinx please provide a list of tests we have done before we are confident enough to merge this PR
> 
> Hi @larry9523 , I have tested xrt flow test on aie2p and vadd test on aie2ps by modifying host application and test passes on both of them.

@larry9523 I have verified non control packet , control packet elf flow tests and also full elf (xclbin2elf) flows and patching args in multiple runs tests passes only after applying this change.
@sayyanna  is verifying preemption test, if that test passes I guess we are good to go.

**Comment by sayyanna** [comment]
> > > @rbramand-xilinx please provide a list of tests we have done before we are confident enough to merge this PR
> > 
> > 
> > Hi @larry9523 , I have tested xrt flow test on aie2p and vadd test on aie2ps by modifying host application and test passes on both of them.
> 
> @larry9523 I have verified non control packet , control packet elf flow tests and also full elf (xclbin2elf) flows and patching args in multiple runs tests passes only after applying this change. @sayyanna is verifying preemption test, if that test passes I guess we are good to go.

Verified preemption test with patching args multiple times and tests passes only after applying this change.

**Review by larry9523 — APPROVED** [review]

**Pull request merged**